### DISCUSSION TITLE
Plugin system

### DISCUSSION THREAD

**Discussion by diegomachadosoares** [TriliumNext/Trilium#1873] [Ideas]
Hi, Guys.

Not sure if this is the right place but I wanted to start a discussion before actually submitting a feature request.
Also, could not find any discussions related.

The thing is, I really like trillium's scripting system and I think that this is what makes it a really powerful system, not just another tool (like many here, I've tried a lot of them!).
In the other hand I think that it lacks a system to "distribute scripts/code/functionality across users" built into the tool (a plugins system, maybe?).

I got to that conclusion after having seen a lot of cool stuff (some I have copied the idea and modified to my own needs) inside issues and, as a side effect, this would also help the ones that cannot code (I also seen a lot of people saying that use Trillium but cannot code).
I is absolutely not a substitution to the scripting system, it is an addition.
Something like a catalog of user scripts. 

So what do you guys think about a "plugin system/scripts catalog" or anything of the sort?

Best regards,

**Comment by chee** [comment]
I think you could build a package manager as a note.

A note that is able to fetch and import other notes from URLs, and checks for updates to those URLs on startup.

I like the distribution mechanism of [trilium timeline](https://gitlab.com/Grymas/trilium-timeline). You hit gitlab's (or github's) built in "download this repo as a zip" button, and then you import that zip into trilium.

**Comment by zadam** [comment]
Hi @diegomachadosoares 

I understand your concerns. As @chee points out it is possible to share the ZIP files, either with "download this repo as a zip" or just the Trilium's exported subtree ZIP file.

But this system has also many weak points:

* there's no centralized repository of available "plugins", discovery is difficult
* scripting API evolves, sometimes in incompatible ways. Exported/imported ZIPs do not have any metadata indicating with which API versions they are compatible
* once you import a ZIP file it's just an ordinary note subtree, and you can do anything with it - configure/modify it
  * but this also means there isn't a good way to upgrade a plugin - if you import again, you will get another copy. You don't know how to copy over your custom modifications (configuration) from the old version

So I don't think current scripting support makes a good easy to use plugin system. I see it as a way for individual advanced users to enhance/extend their experience. You can share your scripts with other advanced users who can dive into scripts but it's probably not a good idea for "normal" users.

Having said that, I would like to see some plugin system. In my thinking it's mainly a packaging layer on top of existing scripts - e.g. package root would have labels defining the package repository, the supported API range. There should be also some delimitation between configuration and the plugin itself. Plugin notes would be read-only by default. Based on this Trilium should be able to execute an upgrade - download the ZIP from the repository defined in the label and replace the plugin notes (not the configuration).

**Reply by TheOneValen** [reply]
The delimitation between configuration and the plugin itself is key. Thats what I struggle with when I want to publish my stuff.

**Comment by meichthys** [comment]
I keep thinking about this every time i see another show and tell discussion. 
For now I've been submitting pull requests for these 'plugins' to: https://github.com/Nriver/awesome-trilium
